go/engine/selfprovision.go

```diff
@@ -0,0 +1,274 @@
+
+	// From this point on, if there's an error, we abort the transaction.
+	defer func() {
+		if tx != nil {
+			tx.Abort()
+		}
+		if err == nil {
+			m = m.CommitProvisionalLogin()
+		}
+	}()
+
+	keys, err := e.loadUserAndActiveDeviceKeys(m)
+	if err != nil {
+		return err
+	}
+
+	// Make new device keys and sign them with current device keys
+	if err = e.provision(m, keys); err != nil {
+		return err
+	}
+
+	// commit the config changes
+	// NOTE we should wrap the Active device change in a transaction as well.
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+
+	// Zero out the TX so that we don't abort it in the defer() exit.
+	tx = nil
+
+	// logs any errors with local storage
+	verifyLocalStorage(m, e.User.GetNormalizedName().String(), e.User.GetUID())
+	if err := e.syncSecretStore(m); err != nil {
+		m.CDebugf("unable to syncSecretStore: %v", err)
+	}
+
+	e.clearCaches(m)
+	e.sendNotification()
+	return nil
+}
+
+func (e *SelfProvisionEngine) loadUserAndActiveDeviceKeys(m libkb.MetaContext) (*libkb.DeviceWithKeys, error) {
+	// run the LoginLoadUser sub-engine to load a user
+	ueng := newLoginLoadUser(e.G(), e.G().Env.GetUsername().String())
+	if err := RunEngine2(m, ueng); err != nil {
+		return nil, err
+	}
+	e.User = ueng.User()
+	pukRing, err := libkb.NewPerUserKeyring(e.G(), e.User.GetUID())
+	if err != nil {
+		return nil, err
+	}
+	e.perUserKeyring = pukRing
+
+	activeDevice := e.G().ActiveDevice
+	encKey, err := activeDevice.EncryptionKey()
+	if err != nil {
+		return nil, err
+	}
+	sigKey, err := activeDevice.SigningKey()
+	if err != nil {
+		return nil, err
+	}
+	keys := libkb.NewDeviceWithKeysOnly(sigKey, encKey)
+	if _, err := keys.Populate(m); err != nil {
+		return nil, err
+	}
+
+	return keys, nil
+}
+
+func (e *SelfProvisionEngine) provision(m libkb.MetaContext, keys *libkb.DeviceWithKeys) error {
+	// Set the active device to be a special provisional key active device,
+	// which keeps a cached copy around for DeviceKeyGen, which requires it to
+	// be in memory.  It also will establish a NIST so that API calls can
+	// proceed on behalf of the user.
+	m = m.WithProvisioningKeyActiveDevice(keys, e.User.ToUserVersion())
+
+	// need lksec to store device keys locally
+	if err := e.fetchLKS(m, keys.EncryptionKey()); err != nil {
+		return err
+	}
+	return e.makeDeviceKeysWithSigner(m, keys.SigningKey())
+}
+
+// copied from loginProvision
+func (e *SelfProvisionEngine) fetchLKS(m libkb.MetaContext, encKey libkb.GenericKey) error {
+	gen, clientLKS, err := fetchLKS(m, encKey)
+	if err != nil {
+		return err
+	}
+	e.lks = libkb.NewLKSecWithClientHalf(clientLKS, gen, e.User.GetUID())
+	return nil
+}
+
+// makeDeviceKeysWithSigner creates device keys given a signing key.
+func (e *SelfProvisionEngine) makeDeviceKeysWithSigner(m libkb.MetaContext, signer libkb.GenericKey) error {
+	if err := e.ensureLKSec(m); err != nil {
+		return err
+	}
+
+	ss, err := m.ActiveDevice().SyncSecrets(m)
```

Comment:
is there a better way to get the device type?

Reply:
upak has it, i believe